pkgs/development/libraries/science/math/cudnn/generic.nix

```diff
@@ -52,7 +52,7 @@ stdenv.mkDerivation {
   meta = with stdenv.lib; {
     description = "NVIDIA CUDA Deep Neural Network library (cuDNN)";
     homepage = "https://developer.nvidia.com/cudnn";
-    license = licenses.unfree;
```

Comment:
Same as the above comment (Sec 1.2)
```
These are the distribution requirements for you to exercise the distribution grant:

    1. Your application must have material additional functionality, beyond the included portions of the SDK.
    2. The distributable portions of the SDK shall only be accessed by your application.
    3. The following notice shall be included in modifications and derivative works of sample source code distributed: “This software contains source code provided by NVIDIA Corporation.”
    4. Unless a developer tool is identified in this Agreement as distributable, it is delivered for your internal use only.
    5. Unless a developer tool is identified in this Agreement as distributable, it is delivered for your internal use only.
    6. You agree to notify NVIDIA in writing of any known or suspected distribution or use of the SDK not in compliance with the requirements of this Agreement, and to enforce the terms of your agreements with respect to distributed SDK.
```
Point 1 and 2 would indicate that it is **not** allowed to redistribute the SDK itself.

Reply:
See section `2.3`

```
2.3. Operating Systems

Those portions of the SDK designed exclusively for use on the Linux or FreeBSD operating systems,
or other operating systems derived from the source code to these operating systems,
may be copied and redistributed for use in accordance with this Agreement,
provided that the object code files are not modified in any way (except for unzipping of compressed files).
```

Reply:
The cuDNN supplement addresses this: `4.2. Distribution. The following portions of the SDK are distributable under the Agreement: the runtime files .so and .h.` This does appear to mean that `libcudnn_static.a` cannot be redistributed, but will reach out to NVIDIA to clarify.

Reply:
Better ask Nvidia, but .a and .so have basically the same use, it's just a matter of static/dynamic linking I believe.

Reply:
Point 1 seems to be a show stopper. A derivation that only contains the redistributable parts of the SDK  would not satisfy 1. The whole closure might though fulfill 1(?) 

Maybe @bhipple or @matthewbauer have an idea here?

Reply:
@markuskowa is it preventing from distributing compiled versions of applications like PyTorch with CUDA? As of today, it does not seem to me like CUDA SDK is the issue, but compilation of applications with CUDA... or am I missing the point?

Reply:
@GuillaumeDesforges The way I read the EULA including CUDA libs in the output of PyTorch is not the problem. However, the way we have it setup the whole CUDA package is stored in its own output. Thus we would redistribute the SDK without any "additional functionality", which the EULA forbids. The point is that we create a nix store path of CUDA that is independent of applications such as PyTorch.

Reply:
Thanks @markuskowa for the clear explanation. Then, would it be possible to keep the current CUDA derivation (and license) and only distribute the subderivations like PyTorch in the cache (Hydra I believe). Then only subderivations need a change in their license.

Reply:
With the organization as it is now that seems not to be possible. Hydra can only build packages that are not marked `unfree`. If PyTorch needs an `unfree` package as input it can't be built by hydra. Moreover,  there is still the problem that fixing `RPATH` in the CUDA libs is formally a violation of the EULA as it modifies the binary.

Reply:
Would it be possible for Hydra to build CUDA, build other derivations that use it, then not distribute CUDA?

Reply:
No, everything that is build by hydra, including the sources, ends up in the cache and is distributed.

Reply:
I reached out to nvidia to clarify if we can distribute the .so only as a nvidia package—this is what conda does. Will post an update when I hear back, but my understanding is this is ok under 2.3 for CUDA, and its unequivocally ok for cuDNN per 4.2.

Reply:
For posterity, here is a record of my outreach to Nvidia. I believe we are ok to proceed, time and manpower permitting

From: Tyler Stephen Benster <[redacted]@stanford.edu>
Sent: Thursday, March 5, 2020 8:49 AM
To: nvidia-compute-license-questions@nvidia.com <nvidia-compute-license-questions@nvidia.com>
Subject: Re: [open science] Redistributing Cuda & cuDNN
 
To whom it may concern,

I wanted to follow up on my previous emails.

My understanding after consulting with others is that what we do in the nixpkgs derivation for CUDA does not preclude binary caching & redistribution, as we only modify the library metadata such as the dynamic section (i.e. DT_RUNPATH, setting interpreter, setting RPATH). As I understand, object code refers to machine code that is executed by the processor, and thus my understanding is that we leave the object code untouched. I previously shared these post-distribution patches on 2/3/2020 for your review.

Furthermore, my understanding under Section 2.3 is that we are ok to redistribute the SDK in full as long as this redistribution only happens under Linux.

Thus, we plan on proceeding with setting up a binary cache for distributing CUDA and packages requiring CUDA using nixpkgs.

Please get in touch if you view any of the above as a violation of the Software License agreement.

Best,

Tyler Benster
Neurosciences PhD Candidate
Stanford University School of Medicine
From: Tyler Stephen Benster
Sent: Monday, February 3, 2020 6:23 PM
To: nvidia-compute-license-questions@nvidia.com <nvidia-compute-license-questions@nvidia.com>
Subject: [open science] Redistributing Cuda & cuDNN
 
Hello,

I'm a scientist at Stanford using CUDA for accelerating neuroscience.

There are a small number of linux-specific patches necessary to use the CUDA libraries using nix, a package manager for reproducible software:

    we have to change the dynamic-linker
    we have to patch the RPATH
    we have to add the opengl run path

These modifications are shown here for CUDA https://github.com/NixOS/nixpkgs/blob/c55809e256024880420bdc101c7ce9562d213dfa/pkgs/development/compilers/cudatoolkit/common.nix and here for cuDNN: https://github.com/NixOS/nixpkgs/blob/1e72fa51580dcac77549ffd9e499db6aa74d3286/pkgs/development/libraries/science/math/cudnn/generic.nix.


I'd like to explicitly ask for permission to distribute software using the CUDA libraries as modified above. This is critical for disseminating my and my colleague's scientific software to the broader community.

Please let me know if this is permissible? Or let me know if I need to contact someone else?

Sincerely,
Tyler

Reply:
Does conda have a special exception from NVIDIA? Their binary patching is every bit as intrusive as what we'd be doing: https://docs.conda.io/projects/conda-build/en/latest/resources/make-relocatable.html#making-packages-relocatable
If they're redistributing patchelf'd binaries, I see no reason why we can't.

Reply:
This is what I found https://github.com/conda-forge/cudatoolkit-feedstock/issues/15#issuecomment-717563755

Reply:
@knedlsepp the comment you mentioned points to https://github.com/conda-forge/cudatoolkit-feedstock/pull/10/files, which supposedly contains a link to an Nvidia bug tracker(?) with a with an approval of the Conda package by Nvidia (http://nvbugs/3052604). However, that link is truncated/garbled. Any guess what the right form of that link is supposed to be?

Reply:
Could be an Nvidia internal tracker. One of the commenters seems to be both Fonda contributor as well as Nvidia employee